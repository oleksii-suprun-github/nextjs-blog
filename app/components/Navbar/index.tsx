'use client';

import { useRef } from 'react';
import Link from 'next/link';
import Image from 'next/image';
import { FiMenu } from 'react-icons/fi';
import Drawer from '@/app/ui-lib/Drawer';
import Navigation from '../Navigation';

const Navbar = () => {
  const drawerToggleRef = useRef<HTMLInputElement>(null);

  const drawerToggleHandler = () => {
    if (drawerToggleRef.current) {
      drawerToggleRef.current.checked = false;
    }
  };

  return (
    <nav className="navbar mb-20 bg-brand-purple px-4 pt-12">
      <div className="container mx-auto">
        <div className="navbar-start">
          <Link href="/">
            <Image src="/logo.svg" alt="Logo" width={150} height={50} priority />
          </Link>
        </div>
        <div className="navbar-center hidden lg:flex">
          <Navigation />
        </div>
        <div className="navbar-end flex">
          <div className="lg:hidden">
            <Drawer toggleRef={drawerToggleRef} icon={<FiMenu size={38} />}>
              <Navigation onClick={drawerToggleHandler} />
              <a className="btn mx-auto w-2/3 bg-brand-pink text-stone-900 hover:bg-brand-light-pink">
                Feedback
              </a>
            </Drawer>
          </div>
          <a className="btn hidden bg-brand-pink text-stone-900 hover:bg-brand-light-pink lg:flex">
            Feedback
          </a>
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
