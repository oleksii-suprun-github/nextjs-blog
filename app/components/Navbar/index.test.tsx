import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import Navbar from '.';

vi.mock('next/link', () => ({
  __esModule: true,
  default: ({ href, children, ...rest }: any) => (
    <a href={href} {...rest}>
      {children}
    </a>
  ),
}));

vi.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    const { src, alt, width, height, ...rest } = props;
    return (
      // eslint-disable-next-line @next/next/no-img-element
      <img src={src} alt={alt} width={width} height={height} {...rest} />
    );
  },
}));

describe('Navbar Component', () => {
  it('should render correctly and match snapshot', () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the logo with correct attributes', () => {
    render(<Navbar />);
    const logo = screen.getByAltText('Logo');
    expect(logo).toBeInTheDocument();
    expect(logo).toHaveAttribute('src', '/logo.svg');
    expect(logo).toHaveAttribute('width', '150');
    expect(logo).toHaveAttribute('height', '50');
  });

  it('should render navigation links in desktop view', () => {
    render(<Navbar />);
    const navLinks = screen.getAllByRole('link');
    expect(navLinks).toHaveLength(9);

    expect(navLinks[0]).toHaveTextContent('');
    expect(navLinks[0]).toHaveAttribute('href', '/');

    expect(navLinks[1]).toHaveTextContent('Home');
    expect(navLinks[1]).toHaveAttribute('href', '/');

    expect(navLinks[2]).toHaveTextContent('About');
    expect(navLinks[2]).toHaveAttribute('href', '#');

    expect(navLinks[3]).toHaveTextContent('Events');
    expect(navLinks[3]).toHaveAttribute('href', '#');

    expect(navLinks[4]).toHaveTextContent('Contact');
    expect(navLinks[4]).toHaveAttribute('href', '#');
  });

  it('should display the Feedback button on desktop view', () => {
    render(<Navbar />);
    const feedbackButton = screen.getByText('Feedback');
    expect(feedbackButton).toBeInTheDocument();
    expect(feedbackButton).toHaveClass(
      'btn hidden bg-brand-pink text-stone-900 hover:bg-brand-light-pink md:flex',
    );
  });
});
