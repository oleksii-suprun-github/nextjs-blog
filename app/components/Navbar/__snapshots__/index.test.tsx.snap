// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar Component > should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar mb-20 bg-brand-purple px-4 pt-12"
  >
    <div
      class="container mx-auto"
    >
      <div
        class="navbar-start"
      >
        <a
          href="/"
        >
          <img
            alt="Logo"
            height="50"
            src="/logo.svg"
            width="150"
          />
        </a>
      </div>
      <div
        class="navbar-center hidden lg:flex"
      >
        <div
          class="mb-12 mt-16 flex flex-col items-center lg:mb-0 lg:mt-0 lg:flex-row lg:items-start"
        >
          <a
            class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
            href="/"
          >
            <span
              class="text-stone-300 hover:text-stone-100"
            >
              Home
            </span>
          </a>
          <a
            class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
            href="#"
          >
            <span
              class="text-stone-300 hover:text-stone-100"
            >
              About
            </span>
          </a>
          <a
            class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
            href="#"
          >
            <span
              class="text-stone-300 hover:text-stone-100"
            >
              Publications
            </span>
          </a>
          <a
            class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
            href="#"
          >
            <span
              class="text-stone-300 hover:text-stone-100"
            >
              Contact
            </span>
          </a>
        </div>
      </div>
      <div
        class="navbar-end flex"
      >
        <div
          class="lg:hidden"
        >
          <div
            class="drawer"
          >
            <input
              class="drawer-toggle"
              id="drawer-mobile"
              type="checkbox"
            />
            <div
              class="drawer-content"
            >
              <label
                class="drawer-button text-stone-300"
                for="drawer-mobile"
              >
                <svg
                  fill="none"
                  height="38"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="38"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="3"
                    x2="21"
                    y1="12"
                    y2="12"
                  />
                  <line
                    x1="3"
                    x2="21"
                    y1="6"
                    y2="6"
                  />
                  <line
                    x1="3"
                    x2="21"
                    y1="18"
                    y2="18"
                  />
                </svg>
              </label>
            </div>
            <div
              class="drawer-side z-30"
            >
              <label
                aria-label="close sidebar"
                class="drawer-overlay"
                for="drawer-mobile"
              />
              <div
                class="menu min-h-full w-80 bg-brand-dark-purple p-4 text-stone-300"
              >
                <div
                  class="mb-12 mt-16 flex flex-col items-center lg:mb-0 lg:mt-0 lg:flex-row lg:items-start"
                >
                  <a
                    class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
                    href="/"
                  >
                    <span
                      class="text-stone-300 hover:text-stone-100"
                    >
                      Home
                    </span>
                  </a>
                  <a
                    class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
                    href="#"
                  >
                    <span
                      class="text-stone-300 hover:text-stone-100"
                    >
                      About
                    </span>
                  </a>
                  <a
                    class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
                    href="#"
                  >
                    <span
                      class="text-stone-300 hover:text-stone-100"
                    >
                      Publications
                    </span>
                  </a>
                  <a
                    class="mb-5 text-lg lg:mb-0 lg:mr-12 lg:text-base"
                    href="#"
                  >
                    <span
                      class="text-stone-300 hover:text-stone-100"
                    >
                      Contact
                    </span>
                  </a>
                </div>
                <a
                  class="btn mx-auto w-2/3 bg-brand-pink text-stone-900 hover:bg-brand-light-pink"
                >
                  Feedback
                </a>
              </div>
            </div>
          </div>
        </div>
        <a
          class="btn hidden bg-brand-pink text-stone-900 hover:bg-brand-light-pink lg:flex"
        >
          Feedback
        </a>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
